src/pubd/repository.rs

```diff
@@ -183,60 +521,78 @@ impl RrdpServer {
+    fn publish(
+        &mut self,
+        elements: DeltaElements,
+        jail: &uri::Rsync,
+        config: &RepositoryRetentionConfig,
+    ) -> KrillResult<()> {
+        if elements.is_empty() {
+            Ok(())
+        } else {
+            self.serial += 1;
 
-        let delta = Delta::new(self.session, next, elements);
+            // Update the snapshot, this can fail if the delta is illegal.
+            self.snapshot.apply_delta(elements.clone(), jail)?;
 
-        let mut next_snapshot = self.snapshot.clone();
-        next_snapshot.apply_delta(delta.clone());
+            self.update_deltas(elements, config);
+            self.update_nofication_file(config);
 
-        let snapshot_uri = self.snapshot_uri(next);
-        let snapshot_path = self.snapshot_path(next);
-        let snapshot_xml = next_snapshot.xml();
-        let snapshot_hash = HexEncodedHash::from_content(snapshot_xml.as_slice());
-        let snapshot_ref = SnapshotRef::new(snapshot_uri, snapshot_path, snapshot_hash);
+            Ok(())
+        }
+    }
 
-        // keep at least 5 deltas if available, but beyond that no
-        // more then the combined size of which would exceed the
-        // size of the snapshot.
-        let snapshot_size = next_snapshot.size();
-        let mut deltas_size = delta.elements().size();
+    // Push the delta and truncate excessive deltas.
+    //  - keep no more than the size of the snapshot
+    //  - don't keep deltas older than configured seconds
+    //  - but keep a minimum of configured deltas
+    fn update_deltas(&mut self, elements: DeltaElements, config: &RepositoryRetentionConfig) {
+        self.deltas.push_front(Delta::new(self.session, self.serial, elements));
+        let mut keep = 0;
+        let mut size = 0;
+        let snapshot_size = self.snapshot.size();
 
-        let mut deltas = vec![&delta];
+        let retain_secs = config.retention_delta_files_seconds;
+        let minimum = config.retention_delta_files_min_nr;
 
         for delta in &self.deltas {
-            if deltas.len() < 5 {
-                deltas.push(delta)
-            } else {
-                deltas_size += delta.elements().size();
-                if deltas_size < snapshot_size {
-                    deltas.push(delta)
-                } else {
-                    break;
-                }
+            size += delta.elements().size();
+            if size > snapshot_size || keep > minimum && delta.older_than_seconds(retain_secs) {
+                break;
             }
+            keep += 1;
         }
+        self.deltas.truncate(keep);
+    }
 
-        let refs: Vec<DeltaRef> = deltas
+    // Update the notification file to include the current snapshot and
+    // deltas. Remove old notification files exceeding the retention time,
+    // so that we can delete old snapshots and deltas which are no longer
+    // relevant.
+    fn update_nofication_file(&mut self, config: &RepositoryRetentionConfig) {
```

Comment:
Does this actually update a file, or just an in memory representation of the notification file that will be written elsewhere to disk?